enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,310 @@
+
+* `logs.app` - Container logs generated by user applications running on the platform, excluding infrastructure container applications
+* `logs.infra` - Logs generated by both infrastructure components running on the platform and OKD nodes (e.g. journal logs).  "Infra" applications are defined as any pods which run in namespaces: `openshift*`, `kube*`, `default`.
+
+Future types may be:
+* `events.kubernetes` - Kubernetes events
+* `audit.container-engine` - Container engine audit events
+
+There are no assumptions regarding whether or not an endpoint is deployed on or off cluster.  Endpoints off-cluster may require adminstrators to perform additional actions in order for logs to be forwarded (e.g. secret creation, opening port) 
+Following is the list of supported endpoint types for this proposal:
+
+* Elasticsearch
+* Fluent secure-forward
+
+### User Stories
+
+#### As an OKD admin, I want to collect logs and send them to my own aggregation solution, for example Splunk
+
+This is a typical example of organizations that desires to re-use their existing enterprise log solution.  We will succeeded if we are able to write logs to their logging service.
+
+#### As an OKD admin, I want to aggregate application logs on-cluster and infra logs off-cluster
+
+This is an example of an OKD cluster hosting solution where several organizations are each provided with a dedicated cluster.  The organization requires access to application container logs but the host requires access to the infra structure logs.
+
+### Implementation Details
+
+#### Assumptions
+
+* A pipeline shall support at least one target endpoint. 
+* A pipeline may support multiple target endpoints.
+* Endpoints that require secrets which are not created by the cluster-logging-operator shall be created and managed by the administrator of the endpoint
+* Secrets shall have keys of: cert,key,cacert which point to the respective certificates for which they represent
+
+#### Security
+* A pipeline may support TLS configuration by declaring a secret. Secrets must contain keys: cert, key, cacert
+* Endpoints defined without a secret will use unsecure communication
+
+#### Cluster Logging With Logstore, Collector, and LogForwarding
+* Undeclared source messages are routed to logstore
+* Declared source messages are routed to the defined service
+#### Cluster Logging With Pipelines Only
+* Undeclared source messages are ignored
+* Declared source messages are routed to the defined service
+
+
+### Risks and Mitigations
+
+## Design Details
+
+The `cluster-logging-operator` will use the forwarding API to generate a collector configuration and mount any secrets into the collector pod.
+
+### API Specification
+The proposed forwarding API:
+
+```
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "ClusterLogging"
+spec:
+  forwarding:
+    targets:
+     - type: “elasticsearch”
+       name: elasticsearch
+       endpoint: elasticsearch.svc.messaging.cluster.local
+       certificates:
```

Comment:
I think this should just be `secretName: name-of-secret` since the secret may contain passwords, shared_keys, and other secure data, other than certs/keys.

Reply:
> I think this should just be `secretName: name-of-secret` since the secret may contain passwords, shared_keys, and other secure data, other than certs/keys.

Per your previous comment, we should restrict the API and limit the config options as much as possible.  The API should be as consistent as possible from one target to the next to facilitate admins modifying the configuration.  Is there a reason we can not be opinionated about some configuration; this is what I chose for the shared key [1].  We are going to have similar issues if we support Kafka.  Can we not be opinionated about the topic to which we write as well?  This seems reasonable as long as we properly document it.  I would prefer to stay away from providing more knobs 

[1] https://github.com/openshift/cluster-logging-operator/pull/231/files#diff-866472c112b52bddfdf28f4ce5ea2413R48

Reply:
All I mean is that it is not obvious/intuitive that I'm supposed to add the `shared_key` value to something under `certificates

Reply:
That is - if there is a field called `certificates` I would expect to put only certificates/keys there.  I would not expect to put the secure_forward `shared_key` field there.  That's why I suggested we get rid of `certificates` and just have a field called `secretName`, where the secret may contain certs, keys, shared_key, passwords, tokens, etc. - basically, anything that we do not want to put as cleartext in the CR.

Reply:
> That is - if there is a field called `certificates` I would expect to put only certificates/keys there. I would not expect to put the secure_forward `shared_key` field there. That's why I suggested we get rid of `certificates` and just have a field called `secretName`, where the secret may contain certs, keys, shared_key, passwords, tokens, etc. - basically, anything that we do not want to put as cleartext in the CR.

Modifying to be `secret` or variant does provide us more freedom to change whats included in the secret in future with a clear understanding that: if key named x exists, the config will look like 'y'.  I would encourage us to remain opinionated about NOT allowing these other types of config into this block and default these to some documented value.